posts/2023-07-24_rounding/rounding.qmd

```diff
@@ -0,0 +1,142 @@
+
+### Bug fix
+
+The first time I needed to round half up in R, I chose the function from a [PHUSE paper](https://www.lexjansen.com/phuse-us/2020/ct/CT05.pdf) and applied it to my study.
+It works fine most of the time, but has issues like the following example -
+
+```{r, message = FALSE}
+# a function that rounds half up
+# exact copy from: https://www.lexjansen.com/phuse-us/2020/ct/CT05.pdf
+ut_round <- function(x, n=0)
+{
+ # x is the value to be rounded
+ # n is the precision of the rounding
+ scale <- 10^n
+ y <- trunc(x * scale + sign(x) * 0.5) / scale
+ # Return the rounded number
+ return(y)
+}
+# round half up for 2436.845, with 2 decimal places
+ut_round(2436.845, 2)
+```
+
+The expected result is 2436.85, but the output rounds it down.
+The function's logic seems fine, I was not able to find the root cause.
+Thanks to the community effort, there are already discussions and resolution available in a [StackOverflow post](https://stackoverflow.com/questions/12688717/round-up-from-5#comment110611119_12688836) -
+
+> There are numerical precision issues, e.g., `round2(2436.845, 2)` returns `2436.84.` Changing `z + 0.5 to z + 0.5 + sqrt(.Machine$double.eps)` seems to work for me. -- Gregor Thomas Jun 24, 2020 at 2:16
+
+After fixing this bug:
+
+```{r, message = FALSE}
+# revised rounds half up
+ut_round1 <- function(x, n=0)
+{
+ # x is the value to be rounded
+ # n is the precision of the rounding
+ scale <- 10^n
+ y <- trunc(x * scale + sign(x) * 0.5 + sqrt(.Machine$double.eps)) / scale
+ # Return the rounded number
+ return(y)
+}
+# round half up for 2436.845, with 2 decimal places
+ut_round1(2436.845, 2)
+```
+
+### We are not alone
+
+The same issue occurred in the following ones as well, and has been raised by users:
+
+-   `janitor::round_half_up()`: [issue](https://github.com/sfirke/janitor/issues/396) was raised and fixed in v2.1.0
+
+-   `Tplyr`: `options(tplyr.IBMRounding = TRUE)`, [issue](https://github.com/atorus-research/Tplyr/issues/124) was raised
+
+-   `scrutiny::round_up_from()/round_up()`: [issue](https://lhdjung.github.io/scrutiny/reference/index.html#rounding) was raised
+
+-   ...
+
+### Which one to use?
+
+The following two functions have the above bug fixed, they both came from the same [StackOverflow post](https://stackoverflow.com/questions/12688717/round-up-from-5).
+
+-   [`janitor::round_half_up()`](https://sfirke.github.io/janitor/reference/round_half_up.html) **version \>= 2.1.0**
+-   [`tidytlg::roundSAS()`](https://pharmaverse.github.io/tidytlg/main/reference/roundSAS.html)
+    -   this function has two more arguments that can convert the result to character and allow a character string to indicate missing values
+
+<!--------------- appendices go here ----------------->
+
+```{r, echo=FALSE}
+source("appendix.R")
+insert_appendix(
+  repo_spec = "pharmaverse/blog",
+  name = long_slug
+)
+```
```

Comment:
This is a minor thing, and perhaps you don't want to do this, but I thought I would mention anyway - would it be worth adding a section looking a the bigger picture, and whether how far we should be trying to go to match SAS rounding? I'm thinking about the fact that if we are trying to move to R-based submissions, then some of this stuff should in a way be expected and accepted. My idea comes from [this comment](https://github.com/pharmaverse/admiral/issues/1872#issuecomment-1523426787) by @rossfarrugia .

Reply:
when moving from SAS to R submission, what kind of discrepancy are expected/accepted for rounding?

Reply:
to add context to my comment before... when using same language for double programming QC often teams would strive for a 100% perfect comparison match, whereas when working in a multi-lingual world maybe it's ok to not perfectly match if differences can be explained in audit-ready QC evidence to explain differences are down to the use of different languages. e.g. a % in first line output with R is 1% and in QC with SAS it is 2%, and upon inspection the QC'er sees that the actual value is 1.5% rounded, hence just add a comment in QC evidence to explain such differences are considered acceptable.

Reply:
>  e.g. a % in first line output with R is 1% and in QC with SAS it is 2%, and upon inspection the QC'er sees that the actual value is 1.5% rounded, hence just add a comment in QC evidence to explain such differences are considered acceptable.

This discrepancy is not caused by SAS/R, but by different rounding approaches, which I think should be based on the study SAP instead of the software. Should the rounding method be clarified first, then choose the corresponding functions in SAS/R?

Reply:
I'd doubt most SAPs would go into such detail as nobody really ever questioned rounding approaches when we only used single language. It's a fair point though. I feel like a closing note such as "Note: with the differences in default behaviour across languages, you could consider your QC strategy and whether an acceptable level of fuzz in the electronic comparisons could be allowed for cases such as rounding when making comparisons between 2 codes written in different languages as long as this is documented. Alternatively you could document the exact rounding approach to be used in the SAP and then match this regardless of programming language used."

Reply:
Good summary, I will add it next week after the holiday. 
The reason I prefer to align with the rounding method during QC is that when we accept the difference, it's easy to overlook the differences caused by other reasons., e.g.,  https://github.com/RConsortium/submissions-pilot3-adam/pull/99

Thanks @rossfarrugia @clarkliming for your comments and discussion!